File: week_04/길찾기게임.py
import sys
sys.setrecursionlimit(10 ** 6)

class Node:
    def __init__(self, value, x):
        self.value = value
        self.x = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.root = None
    
    def insert(self, node):
        if not self.root:
            self.root = node

        current_node = self.root
        while current_node:
            if node.x < current_node.x:
                if current_node.left:
                    current_node = current_node.left
                    continue
                else:
                    current_node.left = node
            if node.x > current_node.x:
                if current_node.right:
                    current_node = current_node.right
                    continue
                else:
                    current_node.right = node
            break
    
    def order(self, pre_order, post_order, node):
        pre_order.append(node.value)
        if node.left:
            self.order(pre_order, post_order, node.left)
        if node.right:
            self.order(pre_order, post_order, node.right)
        post_order.append(node.value)

def solution(nodeinfo):
    nodes = sorted([(i + 1, x, y) for i, (x, y) in enumerate(nodeinfo)], key=lambda node: (-node[2], node[1]))
    
    tree = Tree()
    for value, x, _ in nodes:
        node = Node(value, x)
        tree.insert(node)

    pre_order, post_order = [], []
    tree.order(pre_order, post_order, tree.root)
    return [pre_order, post_order]

File: week_04/길찾기게임.js
class Node {
    constructor(value, x) {
        this.value = value;
        this.x = x;
        this.left = null;
        this.right = null;
    }
}

class Tree {
    constructor() {
        this.root = null;
    }

    insert = (node) => {
        if (!this.root) {
            this.root = node;
            return;
        }

        let currentNode = this.root;
        while (true) {
            if (node.x < currentNode.x) {
                if (!currentNode.left) {
                    currentNode.left = node;
                } else {
                    currentNode = currentNode.left;
                    continue;
                }
            }
            if (node.x > currentNode.x) {
                if (!currentNode.right) {
                    currentNode.right = node;
                } else {
                    currentNode = currentNode.right;
                    continue;
                }
            }
            break;
        }
    };

    order = (answer, node) => {
        answer.preOrder.push(node.value);
        if (node.left) this.order(answer, node.left);
        if (node.right) this.order(answer, node.right);
        answer.postOrder.push(node.value);
    };
}

const compare = (a, b) => {
    if (a[2] == b[2]) {
        return a[1] - b[1];
    }
    return b[2] - a[2];
};

const solution = (nodeinfo) => {
    tree = new Tree();
    nodeinfo
        .map(([x, y], level) => [level + 1, x, y])
        .sort((a, b) => compare(a, b))
        .forEach(([level, x, _]) => {
            node = new Node(level, x);
            tree.insert(node);
        });

    answer = {
        preOrder: [],
        postOrder: [],
    };
    tree.order(answer, tree.root);
    return [answer.preOrder, answer.postOrder];
};
